src/lib/datepicker/datepicker.ts

```diff
@@ -81,23 +85,73 @@ export const MAT_DATEPICKER_SCROLL_STRATEGY_PROVIDER = {
   styleUrls: ['datepicker-content.css'],
   host: {
     'class': 'mat-datepicker-content',
+    '[@tranformPanel]': '"enter"',
     '[class.mat-datepicker-content-touch]': 'datepicker.touchUi',
+    '[class.mat-datepicker-content-above]': '_isAbove',
     '(keydown)': '_handleKeydown($event)',
   },
+  animations: [
+    trigger('tranformPanel', [
+      state('void', style({opacity: 0, transform: 'scale(1, 0)'})),
+      state('enter', style({opacity: 1, transform: 'scale(1, 1)'})),
+      transition('void => enter', animate('400ms cubic-bezier(0.25, 0.8, 0.25, 1)')),
+      transition('* => void', animate('100ms linear', style({opacity: 0})))
+    ]),
+    trigger('fadeInCalendar', [
+      state('void', style({opacity: 0})),
+      state('enter', style({opacity: 1})),
+      transition('void => *', animate('400ms 100ms cubic-bezier(0.55, 0, 0.55, 0.2)'))
+    ])
+  ],
   exportAs: 'matDatepickerContent',
   encapsulation: ViewEncapsulation.None,
   preserveWhitespaces: false,
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class MatDatepickerContent<D> implements AfterContentInit {
-  datepicker: MatDatepicker<D>;
+export class MatDatepickerContent<D> implements AfterContentInit, OnInit, OnDestroy {
+  /** Subscription to changes in the overlay's position. */
+  private _positionChange: Subscription|null;
 
+  /** Reference to the internal calendar component. */
   @ViewChild(MatCalendar) _calendar: MatCalendar<D>;
 
+  /** Reference to the datepicker that created the overlay. */
+  datepicker: MatDatepicker<D>;
+
+  /** Whether the datepicker is above or below the input. */
+  _isAbove: boolean;
+
+  constructor(private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnInit() {
+    if (!this.datepicker._popupRef || this._positionChange) {
+      return;
+    }
+
+    const positionStrategy =
+      this.datepicker._popupRef.getConfig().positionStrategy! as ConnectedPositionStrategy;
+
+    this._positionChange = positionStrategy.onPositionChange.subscribe(change => {
+      const isAbove = change.connectionPair.overlayY === 'bottom';
+
+      if (isAbove !== this._isAbove) {
+        this._isAbove = isAbove;
+        this._changeDetectorRef.markForCheck();
+      }
+    });
+  }
+
   ngAfterContentInit() {
     this._calendar._focusActiveCell();
   }
 
+  ngOnDestroy() {
+    if (this._positionChange) {
```

Comment:
should we use the usual `_destoryed.next()`?

Reply:
I think for this case it's better to got with the subscription, because it allows us to know whether we've subscribed before. Otherwise since we don't dispose of the dialog content until the datepicker is destroyed, we could end up in a situation where we've subscribed multiple times to the same stream.